### PR TITLE
Refactor the legacy way of using pipeline id to create run in KFP backend

### PR DESCRIPTION
**Pull request by jingzhang36**
If users choose to use the legacy way of using pipeline id to create run (instead of using pipeline version in resource references), we substitute it with the pipeline's default version in backend, since pipeline version is now the standard approach to creating a run.

Related issue https://github.com/kubeflow/pipelines/issues/3418

### CODE REVIEW

**Comment by jingzhang36** [comment]
> Thanks! This is really great.
> 
> One behavior discussion:
> Should we pin pipeline version in `Job`?
> I'm not 100% sure, but sounds like if user creates a Job specifying pipeline Id, the user wants the Job to always schedule Runs with latest default pipeline version.
> I believe scheduled workflow controller also calls create run api for runs scheduled by jobs, so it would be enough to just convert the createRun method to pin version.

I thought I did. The Convert*** is also called in CreateJob method in resource_manager.go, and thus the apiJob used to create job now contains a resource reference to the default pipeline version. And then I made the assumption that the runs that got emitted by this job will use the resource reference specified in Job.

**Comment by Bobgy** [comment]
I was more thinking from api interface perspective, if I deploy a Job that only specifies pipeline ID, my expectation is that the Job will schedule each pipeline run using the default version (latest version) at the time runs are scheduled. Rather than the default version when the Job was created.

**Comment by jingzhang36** [comment]
> I was more thinking from api interface perspective, if I deploy a Job that only specifies pipeline ID, my expectation is that the Job will schedule each pipeline run using the default version (latest version) at the time runs are scheduled. Rather than the default version when the Job was created.
> 
> So it seems to me it's more reasonable to not convert Job's pipline ID to version reference.

I see what you mean now. I would prefer have a same version for all the following runs though. Say, (1) if a job is run periodically to pick the change in the input data, you would probably prefer to mix processing logic change with input data change in the final resulting output. (2) before the version is introduced, all the runs created from a job uses the same manifest. So i would probably choose to use the manifest (version) at the job creation time and use this same one for all the recurring runs.

File: backend/src/apiserver/resource/resource_manager_util.go [hunk 1]
```diff
@@ -228,3 +229,28 @@ func OverrideParameterWithSystemDefault(workflow util.Workflow, apiRun *api.Run)
 	}
 	return nil
 }
+
+// Convert PipelineId in PipelineSpec to the pipeline's default pipeline version.
+// This is for legacy usage of pipeline id to create run. The standard way to
+// create run is by specifying the pipeline version.
```
Comment (rmgogogo) [inline backend/src/apiserver/resource/resource_manager_util.go:235]: you mean to Latest Pipeline Version, right?

**Comment by rmgogogo** [comment]
Thanks Jing. As long as it means execute latest version, feel free to submit after Yuan's lgtm without mine.

/retest

**Comment by Bobgy** [comment]
> > I was more thinking from api interface perspective, if I deploy a Job that only specifies pipeline ID, my expectation is that the Job will schedule each pipeline run using the default version (latest version) at the time runs are scheduled. Rather than the default version when the Job was created.
> > So it seems to me it's more reasonable to not convert Job's pipline ID to version reference.
> 
> I see what you mean now. I would prefer have a same version for all the following runs though. Say, (1) if a job is run periodically to pick the change in the input data, you would probably prefer to mix processing logic change with input data change in the final resulting output. (2) before the version is introduced, all the runs created from a job uses the same manifest. So i would probably choose to use the manifest (version) at the job creation time and use this same one for all the recurring runs.

Can we have a chat on this?

**Comment by Bobgy** [comment]
discussed over chat
https://github.com/Bobgy/pipelines/blob/3bf7c0f678f4de71d8b857dc48822083b063856f/backend/src/crd/controller/scheduledworkflow/controller.go#L499

Currently, scheduled workflow controller creates workflows directly. So it cannot keep using latest pipeline versions. We probably need to revisit this to support it, but we will wait for more user interest.

File: backend/src/apiserver/resource/resource_manager_util.go (hunk 1 above)
Comment (jingzhang36) [reply backend/src/apiserver/resource/resource_manager_util.go:235]: It's actually default version. Usually the default is the latest. But we reserve the possibility to add feature such as allowing users to set a default version.

**Pull request merged**